### PR TITLE
feat: Write out a sample `.env` file

### LINKED ISSUE #881: Make a command to write out a sample `.env` instead of having docs/env.example
Split out from https://github.com/influxdata/influxdb_iox/issues/599, related to https://github.com/influxdata/influxdb_iox/pull/608 and https://github.com/influxdata/influxdb_iox/pull/640.

Especially [this comment](https://github.com/influxdata/influxdb_iox/pull/640#issuecomment-758609466):

> Regarding the config command what you suggest sounds perfect - it can spit out a .env file and should all will work nicely. I've fixed the .env sourcing so it shows the values in the help text now too.

Echoed [over here](https://github.com/influxdata/influxdb_iox/pull/834#issuecomment-782125034):

> I agree that getting out of sync could be an issue, but a sample file is always very welcome when I first start using a project - perhaps there's a way to generate a sample file from the config struct? Something like influxdb_iox config dump ?

So this issue is specifically to make some command that will print out documentation, environment variable names, and their current values to stdout in a way that the command could be run with `> .env` and have a working config file to start tweaking.

### PR DESCRIPTION
**Pull request by shepmaster**
Closes #881

<details>
<summary>Example output</summary>

```
# IOX_ADDR=http://127.0.0.1:8082

# This controls the IOx server logging level, as described in https://crates.io/crates/env_logger.
# 
# Levels for different modules can be specified as well. For example `debug,hyper::proto::h1=info` specifies debug logging for all modules except for the `hyper::proto::h1' module which will only display info level logging.
# RUST_LOG=

# Log message format. Can be one of:
# 
# "rust" (default) "logfmt" (logfmt/Heroku style - https://brandur.org/logfmt)
# INFLUXDB_IOX_LOG_FORMAT=

# The identifier for the server.
# 
# Used for writing to object storage and as an identifier that is added to replicated writes, WAL segments and Chunks. Must be unique in a group of connected or semi-connected IOx servers. Must be a number that can be represented by a 32-bit unsigned integer.
# INFLUXDB_IOX_ID=

# INFLUXDB_IOX_BIND_ADDR=127.0.0.1:8080

# INFLUXDB_IOX_GRPC_BIND_ADDR=127.0.0.1:8082

# INFLUXDB_IOX_DB_DIR=

# Which object storage to use. If not specified, defaults to memory.
# 
# Possible values (case insensitive):
# 
# * memory (default): Effectively no object persistence.
# * file: Stores objects in the local filesystem. Must also set `--data-dir`.
# * s3: Amazon S3. Must also set `--bucket`, `--aws-access-key-id`, `--aws-secret-access-key`, and
#    possibly `--aws-default-region`.
# * google: Google Cloud Storage. Must also set `--bucket` and `--google-service-account`.
# * azure: Microsoft Azure blob storage. Must also set `--bucket`, `--azure-storage-account`,
#    and `--azure-storage-access-key`.
#         
# Possible values: memory, file, s3, google, azure
# INFLUXDB_IOX_OBJECT_STORE=

# Name of the bucket to use for the object store. Must also set `--object-store` to a cloud object storage to have any effect.
# 
# If using Google Cloud Storage for the object store, this item as well as `--google-service-account` must be set.
# 
# If using S3 for the object store, must set this item as well as `--aws-access-key-id` and `--aws-secret-access-key`. Can also set `--aws-default-region` if not using the fallback region.
# 
# If using Azure for the object store, set this item to the name of a container you've created in the associated storage account, under Blob Service > Containers. Must also set `--azure-storage-account` and `--azure-storage-access-key`.
# INFLUXDB_IOX_BUCKET=

# When using Amazon S3 as the object store, set this to an access key that has permission to read from and write to the specified S3 bucket.
# 
# Must also set `--object-store=s3`, `--bucket`, and `--aws-secret-access-key`. Can also set `--aws-default-region` if not using the fallback region.
# 
# Prefer the environment variable over the command line flag in shared environments.
# AWS_ACCESS_KEY_ID=

# When using Amazon S3 as the object store, set this to the secret access key that goes with the specified access key ID.
# 
# Must also set `--object-store=s3`, `--bucket`, `--aws-access-key-id`. Can also set `--aws-default-region` if not using the fallback region.
# 
# Prefer the environment variable over the command line flag in shared environments.
# AWS_SECRET_ACCESS_KEY=

# When using Amazon S3 as the object store, set this to the region that goes with the specified bucket if different from the fallback value.
# 
# Must also set `--object-store=s3`, `--bucket`, `--aws-access-key-id`, and `--aws-secret-access-key`.
# AWS_DEFAULT_REGION=us-east-1

# When using Google Cloud Storage as the object store, set this to the path to the JSON file that contains the Google credentials.
# 
# Must also set `--object-store=google` and `--bucket`.
# GOOGLE_SERVICE_ACCOUNT=

# When using Microsoft Azure as the object store, set this to the name you see when going to All Services > Storage accounts > [name].
# 
# Must also set `--object-store=azure`, `--bucket`, and `--azure-storage-access-key`.
# AZURE_STORAGE_ACCOUNT=

# When using Microsoft Azure as the object store, set this to one of the Key values in the Storage account's Settings > Access keys.
# 
# Must also set `--object-store=azure`, `--bucket`, and `--azure-storage-account`.
# 
# Prefer the environment variable over the command line flag in shared environments.
# AZURE_STORAGE_ACCESS_KEY=

# If set, Jaeger traces are emitted to this host using the OpenTelemetry tracer.
# 
# NOTE: The OpenTelemetry agent CAN ONLY be configured using environment variables. It CAN NOT be configured using the command line at this time. Some useful variables:
# 
# * OTEL_SERVICE_NAME: emitter service name (iox by default) * OTEL_EXPORTER_JAEGER_AGENT_HOST: hostname/address of the collector * OTEL_EXPORTER_JAEGER_AGENT_PORT: listening port of the collector.
# 
# The entire list of variables can be found in https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-environment-variables.md#jaeger-exporter
# OTEL_EXPORTER_JAEGER_AGENT_HOST=
```

</details>

### CODE REVIEW

File: src/commands/database.rs [hunk 1]
```diff
@@ -1,8 +1,8 @@
+use clap::Clap;
 use influxdb_iox_client::{
     connection::Builder,
     management::{generated_types::*, *},
 };
-use structopt::StructOpt;
```
Comment (alamb) [inline src/commands/database.rs:5]: Using something built into Clap is cool -- FYI @tustvold

File: Cargo.toml [hunk 2]
```diff
@@ -59,7 +59,7 @@ wal = { path = "wal" }
 byteorder = "1.3.4"
 bytes = "1.0"
 chrono = "0.4"
-clap = "2.33.1"
+clap = { git = "https://github.com/clap-rs/clap", rev = "09009761ec5f6be4d406f70042d85fca543eeca8" }
```
Comment (alamb) [inline Cargo.toml:62]: Can you please add a comment about the rationale for not using a released version of clap?

Something perhaps like:

```suggestion
# Use Clap 3 beta to get structopt support for auto generating .env
# https://github.com/influxdata/influxdb_iox/issues/937: update to released clap3 when available
clap = { git = "https://github.com/clap-rs/clap", rev = "09009761ec5f6be4d406f70042d85fca543eeca8" }
```

If that is correct,

File: src/main.rs [hunk 3]
```diff
@@ -246,6 +249,46 @@ fn get_runtime(num_threads: Option<usize>) -> Result<Runtime, std::io::Error> {
     }
 }
 
+fn dotenv_skeleton() {
```
Comment (alamb) [inline src/main.rs:252]:
```suggestion
/// Print a `.env` file to stdout which contains all environment
/// variables used to control IOx by introspecting the
/// Clap arguments
fn dotenv_skeleton() {
```

File: src/main.rs [hunk 4]
```diff
@@ -113,13 +113,15 @@ enum Command {
     // Clippy recommended boxing this variant because it's much larger than the others
     Server(Box<commands::server::Config>),
     Writer(commands::writer::Config),
+    /// Print a template .env file to standard out
```
Comment (alamb) [inline src/main.rs:116]:
```suggestion
    /// Print a template .env file to standard out
    /// describing all configuration environment
   /// variables.
```

**Review by alamb — APPROVED** [review]
This is very cool @shepmaster  -- thank you! I tested it out:

```
(arrow_dev) alamb@MacBook-Pro:~/Software/influxdb_iox2$ ./target/debug/influxdb_iox help
influxdb_iox 

InfluxDB IOx server and command line tools

USAGE:
    influxdb_iox [FLAGS] [OPTIONS] [SUBCOMMAND]

FLAGS:
        --help       Prints help information
    -v, --verbose    
    -V, --version    Prints version information

OPTIONS:
    -h, --host <host>                  gRPC address of IOx server to connect to [env: IOX_ADDR=]
                                       [default: http://127.0.0.1:8082]
        --num-threads <num-threads>    Set the maximum number of threads to use. Defaults to the
                                       number of cores on the system

SUBCOMMANDS:
    convert            Convert one storage format to another
    database           Manage IOx databases
    dotenv-skeleton    Print a template .env file to standard out
    help               Prints this message or the help of the given subcommand(s)
    meta               Print out metadata information about a storage file
    server             Runs in server mode (default)
    stats              Print out storage statistics information to stdout
    writer             Manage IOx writer ID
```

```
(arrow_dev) alamb@MacBook-Pro:~/Software/influxdb_iox2$ ./target/debug/influxdb_iox dotenv-skeleton
# IOX_ADDR=http://127.0.0.1:8082

# This controls the IOx server logging level, as described in https://crates.io/crates/env_logger.
# 
# Levels for different modules can be specified as well. For example `debug,hyper::proto::h1=info` specifies debug logging for all modules except for the `hyper::proto::h1' module which will only display info level logging.
# RUST_LOG=

# Log message format. Can be one of:
# 
# "rust" (default) "logfmt" (logfmt/Heroku style - https://brandur.org/logfmt)
# INFLUXDB_IOX_LOG_FORMAT=
...
```

I also tested that things seemed to be copacetic when starting the server using the auto generated .env

```
(arrow_dev) alamb@MacBook-Pro:~/Software/influxdb_iox2$  ./target/debug/influxdb_iox dotenv-skeleton > .env
(arrow_dev) alamb@MacBook-Pro:~/Software/influxdb_iox2$ ./target/debug/influxdb_iox
Mar 06 05:50:06.642  WARN influxdb_iox::influxdb_ioxd: NO PERSISTENCE: using Memory for object storage
Mar 06 05:50:06.643  WARN influxdb_iox::influxdb_ioxd: server ID not set. ID must be set via the INFLUXDB_IOX_ID config or API before writing or querying data.
Mar 06 05:50:06.643  INFO influxdb_iox::influxdb_ioxd: gRPC server listening bind_address=127.0.0.1:8082
Mar 06 05:50:06.654  INFO influxdb_iox::influxdb_ioxd: HTTP server listening bind_address=127.0.0.1:8080
Mar 06 05:50:06.654  INFO influxdb_iox::influxdb_ioxd: InfluxDB IOx server ready git_hash="878f293f7f98dfccb81b2648b57cbbaba7527b95"
^C
```

I had some suggestions on comments / wording tweaks but the code looks really nice. 👍

File: src/commands/database.rs (hunk 1 above)
Comment (tustvold) [reply src/commands/database.rs:5]: Structopt is merged into clap 3, it's the same proc macro under the hood if I'm not mistaken

File: Cargo.toml (hunk 2 above)
Comment (tustvold) [reply Cargo.toml:62]: So I actually had a PR that made this migration and I pulled it as the clap 3 API is not yet stable. In particular see [here](https://github.com/clap-rs/clap/discussions/2379)

File: src/commands/database.rs (hunk 1 above)
Comment (shepmaster) [reply src/commands/database.rs:5]: Yep. From my commit message:  

> This merges structopt into Clap proper

**Comment by shepmaster** [comment]
> I'm therefore a little bit apprehensive about this change...

Based on this feedback, I'm closing the PR. It can be reopened / updated when circumstances change.

**Pull request closed without merging**